strings/base_error.h

```diff
@@ -546,6 +546,28 @@ namespace winrt::impl
         }
         return result;
     }
+
+    template <typename T>
+    WINRT_IMPL_NOINLINE void check_cast_result(T* from WINRT_IMPL_SOURCE_LOCATION_ARGS)
+    {
+        if (!from)
+        {
+            com_ptr<impl::IRestrictedErrorInfo> restrictedError;
+            if (WINRT_IMPL_GetRestrictedErrorInfo(restrictedError.put_void()) == 0)
+            {
+                WINRT_IMPL_SetRestrictedErrorInfo(restrictedError.get());
+
+                int32_t code;
+                impl::bstr_handle description;
+                impl::bstr_handle restrictedDescription;
+                impl::bstr_handle capabilitySid;
+                if (restrictedError->GetErrorDetails(description.put(), &code, restrictedDescription.put(), capabilitySid.put()) == 0)
+                {
+                    throw hresult_error(code, take_ownership_from_abi WINRT_IMPL_SOURCE_LOCATION_FORWARD);
+                }
+            }
+        }
+    }
```

Comment:
I'm wondering if code gen would be better as
```cpp
WINRT_IMPL_NOINLINE [[noreturn]] void throw_failed_cast(hresult code WINRT_IMPL_SOURCE_LOCATION_ARGS)
{
    throw hresult_error(code, take_ownership_from_abi WINRT_IMPL_SOURCE_LOCATION_FORWARD);
}

inline void check_cast_result(hresult code WINRT_IMPL_SOURCE_LOCATION_ARGS)
{
    if (code < 0) throw_failed_cast(code WINRT_IMPL_SOURCE_LOCATION_FORWARD);
}
```
where we have a new try_as_reason that returns both a pointer and an hresult.
```cpp
void* result{};
hresult code = ptr->QueryInterface(guid_of<To>(), &result);
return { wrap_as_result<To>(result), code };
```